File: POLIMORFISMO/EJ_9/EJ_9.py
class Bloque:
    def __init__(self):
        self.tipo = "Bloque genérico"
    
    def accion(self):
        print("Realizando acción genérica de bloque")
    
    def colocar(self, orientacion="suelo"):
        print(f"Colocando bloque en el {orientacion}")
    
    def romper(self):
        print("Rompiendo bloque genérico")

class BloqueCofre(Bloque):
    def __init__(self, capacidad, resistencia):
        super().__init__()
        self.capacidad = capacidad
        self.resistencia = resistencia
        self.tipo = "Cofre"
    
    def accion(self):
        print(f"Abriendo cofre de {self.capacidad} slots")
    
    def romper(self):
        print("Rompiendo cofre... ¡Cuidado con los items que caen!")

class BloqueTnt(Bloque):
    def __init__(self, tipo_explosivo, danio):
        super().__init__()
        self.tipo_explosivo = tipo_explosivo
        self.danio = danio
        self.tipo = "TNT"
    
    def accion(self):
        print(f"¡La TNT de tipo {self.tipo_explosivo} está a punto de explotar!")
    
    def romper(self):
        print(f"¡BOOM! La TNT explotó causando {self.danio} de daño")

class BloqueHorno(Bloque):
    def __init__(self, color, capacidad_comida):
        super().__init__()
        self.color = color
        self.capacidad_comida = capacidad_comida
        self.tipo = "Horno"
    
    def accion(self):
        print(f"Cocinando comida en el horno {self.color}")
    
    def romper(self):
        print("Rompiendo horno... ¡Recuerda recoger la comida cocinada!")

# Instanciación de bloques
cofre_madera = BloqueCofre(27, 10)
cofre_ender = BloqueCofre(54, 30)

tnt_normal = BloqueTnt("Normal", 10)
tnt_super = BloqueTnt("Super", 20)

horno_piedra = BloqueHorno("Gris", 3)
horno_nether = BloqueHorno("Negro", 5)

# Ejemplos de uso
cofre_madera.accion()
cofre_madera.colocar()
cofre_madera.romper()

tnt_normal.accion()
tnt_normal.colocar("pared")
tnt_normal.romper()

horno_piedra.accion()
horno_piedra.colocar("techo")
horno_piedra.romper()

File: POLIMORFISMO/EJ_9/EJ_9.java
class Bloque {
    protected String tipo;
    
    public void accion() {
        System.out.println("Realizando acción genérica de bloque");
    }
    
    public void colocar() {
        System.out.println("Colocando bloque en el suelo");
    }
    
    public void colocar(String orientacion) {
        System.out.println("Colocando bloque en la " + orientacion);
    }
    
    public void romper() {
        System.out.println("Rompiendo bloque genérico");
    }
}

class BloqueCofre extends Bloque {
    private int capacidad;
    private int resistencia;
    
    public BloqueCofre(int capacidad, int resistencia) {
        this.capacidad = capacidad;
        this.resistencia = resistencia;
        this.tipo = "Cofre";
    }
    
    @Override
    public void accion() {
        System.out.println("Abriendo cofre de " + capacidad + " slots");
    }
    
    @Override
    public void romper() {
        System.out.println("Rompiendo cofre... ¡Cuidado con los items que caen!");
    }
}

class BloqueTnt extends Bloque {
    private String tipoExplosivo;
    private int danio;
    
    public BloqueTnt(String tipoExplosivo, int danio) {
        this.tipoExplosivo = tipoExplosivo;
        this.danio = danio;
        this.tipo = "TNT";
    }
    
    @Override
    public void accion() {
        System.out.println("¡La TNT de tipo " + tipoExplosivo + " está a punto de explotar!");
    }
    
    @Override
    public void romper() {
        System.out.println("¡BOOM! La TNT explotó causando " + danio + " de daño");
    }
}

class BloqueHorno extends Bloque {
    private String color;
    private int capacidadComida;
    
    public BloqueHorno(String color, int capacidadComida) {
        this.color = color;
        this.capacidadComida = capacidadComida;
        this.tipo = "Horno";
    }
    
    @Override
    public void accion() {
        System.out.println("Cocinando comida en el horno " + color);
    }
    
    @Override
    public void romper() {
        System.out.println("Rompiendo horno... ¡Recuerda recoger la comida cocinada!");
    }
}

public class Minecraft {
    public static void main(String[] args) {
        BloqueCofre cofreMadera = new BloqueCofre(27, 10);
        BloqueCofre cofreEnder = new BloqueCofre(54, 30);
        
        BloqueTnt tntNormal = new BloqueTnt("Normal", 10);
        BloqueTnt tntSuper = new BloqueTnt("Super", 20);
        
        BloqueHorno hornoPiedra = new BloqueHorno("Gris", 3);
        BloqueHorno hornoNether = new BloqueHorno("Negro", 5);
        
        cofreMadera.accion();
        cofreMadera.colocar();
        cofreMadera.romper();
        
        tntNormal.accion();
        tntNormal.colocar("pared");
        tntNormal.romper();
        
        hornoPiedra.accion();
        hornoPiedra.colocar("techo");
        hornoPiedra.romper();
    }
}